### PR TITLE
Multiple dispatch compatibility for copy_ in XLA.

### LINKED ISSUE #1006: PT_BREAK: Implement multiple dispatch
https://github.com/pytorch/pytorch/pull/25653

### PR DESCRIPTION
**Pull request by ezyang**
Basically, `self` may not be an XLA tensor. So we must test if it is one. If it's a CPU tensor, use the `copy_from_` logic (dst => self, self => src)

Must land at the same time as https://github.com/pytorch/pytorch/pull/25653

fixes #1006 

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: torch_xla/csrc/aten_xla_type.cpp [hunk 1]
```diff
@@ -866,12 +866,26 @@ AtenXlaType::convolution_backward_overrideable(
...
 at::Tensor& AtenXlaType::copy_(at::Tensor& self, const at::Tensor& src,
                                bool non_blocking) {
-  XLATensor self_tensor = bridge::GetXlaTensor(self);
+  c10::optional<XLATensor> self_tensor = bridge::TryGetXlaTensor(self);
   c10::optional<XLATensor> src_tensor = bridge::TryGetXlaTensor(src);
-  if (src_tensor) {
-    XLATensor::copy_(self_tensor, *src_tensor);
+
+  if (!src_tensor) {
+    TORCH_ASSERT(self_tensor);
```
Comment (dlibenzi) [inline torch_xla/csrc/aten_xla_type.cpp]: ```XLA_CHECK(self_tensor)```

File: torch_xla/csrc/aten_xla_type.cpp [hunk 2]
```diff
@@ -866,12 +866,26 @@ AtenXlaType::convolution_backward_overrideable(
...
+    self_tensor.SetTensor(CopyTensor(src, self_tensor->scalar_type()));
+  } else if (!self_tensor) {
+    // TODO: Is self_tensor good enough?  I don't think so... therefore
+    // the hack below:
+    //
+    // Do not mark the tensor creation as writeable to not discard the XLA tensor
```
Comment (dlibenzi) [inline torch_xla/csrc/aten_xla_type.cpp]: Comment no more valid.

Comment (ezyang) [reply torch_xla/csrc/aten_xla_type.cpp]: I can delete the comment. However, I am curious to know if I need to do all this faffing around in the first place. Ideally, I want to write the code like this:

```
  if (!src_tensor) {
    XLA_CHECK(self_tensor);
    self_tensor.SetTensor(CopyTensor(src, self_tensor->scalar_type()));
  } else if (!self_tensor) {
    at::Tensor t = CopyTensor(*self_tensor, self.scalar_type());
    const_cast<at::Tensor&>(self).unsafeGetTensorImpl()->shallow_copy_from(
        t.getIntrusivePtr());
```

i.e., not going through the XLA tensor list at all. Is that sound?

Comment (dlibenzi) [reply torch_xla/csrc/aten_xla_type.cpp]: > I can delete the comment. However, I am curious to know if I need to do all this faffing around in the first place. Ideally, I want to write the code like this:
> 
> ```
>   if (!src_tensor) {
>     XLA_CHECK(self_tensor);
>     self_tensor.SetTensor(CopyTensor(src, self_tensor->scalar_type()));
>   } else if (!self_tensor) {
>     at::Tensor t = CopyTensor(*self_tensor, self.scalar_type());
>     const_cast<at::Tensor&>(self).unsafeGetTensorImpl()->shallow_copy_from(
>         t.getIntrusivePtr());
> ```
> 
> i.e., not going through the XLA tensor list at all. Is that sound?

Wait, what? :grin:
You do need to get a CPU tensor our of the source (XLA) tensor.
How the above code seven work?
It seems to be copying self into self...

Comment (ezyang) [reply torch_xla/csrc/aten_xla_type.cpp]: I typoed!

```
  if (!src_tensor) {
    XLA_CHECK(self_tensor);
    self_tensor.SetTensor(CopyTensor(src, self_tensor->scalar_type()));
  } else if (!self_tensor) {
    at::Tensor t = CopyTensor(*src, self.scalar_type());
    const_cast<at::Tensor&>(self).unsafeGetTensorImpl()->shallow_copy_from(
        t.getIntrusivePtr());
```

**Comment by ezyang** [comment]
Absolutely. The diff is here just so I can test XLA on CI here. We can remove it before landing.

**Pull request merged**